### PR TITLE
Add Mike's module to dump raw waveform for ROI finding network training.

### PR DESCRIPTION
**Pull request by yangtj207**

### CODE REVIEW

**Comment by lgarren** [comment]
code-checks

**Comment by lgarren** [comment]
+code-checks

I have confirmed that the whitespace problem was fixed, but is not being reported properly.

**Comment by lgarren** [comment]
trigger build

File: larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+  fSelectProcID   = p.get<std::string>("SelectProcID",  "ANY");
+  fSelectPDGCode  = p.get<int>("SelectPDGCode", 0);
+
+  fPlaneToDump = p.get<std::string>("PlaneToDump", "U");
+  fMinParticleEnergyGeV = p.get<double>("MinParticleEnergyGeV",0.);
+  fMinEnergyDepositedMeV = p.get<double>("MinEnergyDepositedMeV",0.);
+  fMinNumberOfElectrons = p.get<int>("MinNumberOfElectrons",1000);
+  fMaxNumberOfElectrons = p.get<int>("MaxNumberOfElectrons",100000);
+
+  fSaveSignal = p.get<bool>("SaveSignal", true);
```
Comment (absolution1) [inline larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc:171]: These variables could be constructed in the initialisation list

`nnet::RawWaveformDump::RawWaveformDump(fhicl::ParameterSet const& p)
   : EDAnalyzer{p},
     fSimulationProducerLabel(p.get<std::string>("SimulationProducerLabel", "largeant")),
.....
     `

File: larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+  std::string fPlaneToDump;
+  double fMinParticleEnergyGeV;
+  double fMinEnergyDepositedMeV;
+  int fMinNumberOfElectrons;
+  int fMaxNumberOfElectrons;
+  bool fSaveSignal;
+  art::ServiceHandle<geo::Geometry> fgeom;
+  art::ServiceHandle<cheat::ParticleInventoryService> PIS;
+  //art::ServiceHandle<SimChannelExtractService> m_pscx;
+  detinfo::DetectorClocks const * fClks;
```
Comment (absolution1) [inline larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc:105]: Are any of these needed as data members or could they be created on the fly as they are all services?

File: larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+  };
+  std::string get_gen(int tid){
+    if( !isSorted ){
+      this->sort_now();
+    }
+    return std::lower_bound(track_id_map.begin(), track_id_map.end(), tid,[](const auto &a, const auto &b){return (a.first < b) ; } )->second;
+  };
+
+};
+genFinder* gf = new genFinder();
```
Comment (absolution1) [inline larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc:141]: Can this be wrapped in a smart pointer?

File: larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+  return;
+}
+
+//-----------------------------------------------------------------------
+void nnet::RawWaveformDump::beginJob()
+{
+
+  auto const* detprop = lar::providerFrom<detinfo::DetectorPropertiesService>();
+
+  std::random_device rndm_device;	// this will give us our seed
```
Comment (absolution1) [inline larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc:182]: Would it be preferable to use the art random seed service here?  The randomness would, in principle, be reproducible.

File: larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc (hunk 2 above)
Comment (yangtj207) [reply larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc:105]: No, it is not needed but there is nothing wrong with this implementation.

File: larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc (hunk 1 above)
Comment (yangtj207) [reply larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc]: This is now addressed.

File: larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc (hunk 3 above)
Comment (yangtj207) [reply larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc]: Changed as suggested.

File: larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc (hunk 4 above)
Comment (yangtj207) [reply larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc:172]: This module produces samples for CNN training. Reproducibility is not needed.

**Comment by lgarren** [comment]
trigger build

There are no whitespace problems.  We still don't know why the code checks fail.

**Comment by knoepfel** [comment]
@yangtj207, what is the `c2numpy` utility used for?

**Comment by yangtj207** [comment]
> @yangtj207, what is the `c2numpy` utility used for?

@mhlswang should also comment. 
My understanding is that it helps to save information to numpy format.

**Comment by knoepfel** [comment]
@yangtj207, thanks.  Before I make specific comments on the code, I think there needs to be a discussion on whether `c2numpy` should be incorporated.  It was written 3.5 years ago (https://github.com/diana-hep/c2numpy), and it has been archived, meaning that the intention is for it to not be maintained.  Also, because the instructions for using it say to just embed the header file in your project, it is clear that LArSoft would be taking on a support burden.  

Another alternative that might do something similar and is under active development is https://github.com/dpilger26/NumCpp.  Can you and @mhlswang take a look to determine if that is an adequate solution?

**Comment by yangtj207** [comment]
> @yangtj207, thanks. Before I make specific comments on the code, I think there needs to be a discussion on whether `c2numpy` should be incorporated. It was written 3.5 years ago (https://github.com/diana-hep/c2numpy), and it has been archived, meaning that the intention is for it to not be maintained. Also, because the instructions for using it say to just embed the header file in your project, it is clear that LArSoft would be taking on a support burden.
> 
> Another alternative that might do something similar and is under active development is https://github.com/dpilger26/NumCpp. Can you and @mhlswang take a look to determine if that is an adequate solution?

Thank you @knoepfel  for your comments. I looked into NumCpp and it does not seem trivial to convert the code to use NumCpp. Mike and I had a little discussion on this. Even though the code is potentially useful to multiple experiments, we understand your concern with maintaining
c2numpy. Therefore, if you agree, we would like to delete this pull request. Experiments that wish to use this code can implement it in their own repository. Thanks.

**Comment by aihimmel** [comment]
Moving a potentially useful tool out of the common framework seems like a step in the wrong direction. Can we include it, and if the include stops working, cross that bridge when we get there? Otherwise we may need to duplicate this work across multiple collaborations, rather than having 1 collaboration fix it for everyone.


On Apr 29, 2020, at 6:57 AM, Tingjun Yang <notifications@github.com> wrote:

﻿

@yangtj207<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_yangtj207&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=ILbuN39K86TRVV4M0fXV7A&m=WTxe_jLFiZc19hQ0WBjz01ksdtPAwYEvCZEUs92JSJ0&s=y3IAycXmRX76DDSTN1fuUArIMo0bgmOeFFiWPosCvJg&e=>, thanks. Before I make specific comments on the code, I think there needs to be a discussion on whether c2numpy should be incorporated. It was written 3.5 years ago (https://github.com/diana-hep/c2numpy<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_diana-2Dhep_c2numpy&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=ILbuN39K86TRVV4M0fXV7A&m=WTxe_jLFiZc19hQ0WBjz01ksdtPAwYEvCZEUs92JSJ0&s=l_zu9gaf0ayoF-yUjTF7bc9wd3oSeoEo2qqgbLqVD_s&e=>), and it has been archived, meaning that the intention is for it to not be maintained. Also, because the instructions for using it say to just embed the header file in your project, it is clear that LArSoft would be taking on a support burden.

Another alternative that might do something similar and is under active development is https://github.com/dpilger26/NumCpp<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dpilger26_NumCpp&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=ILbuN39K86TRVV4M0fXV7A&m=WTxe_jLFiZc19hQ0WBjz01ksdtPAwYEvCZEUs92JSJ0&s=pMcwdfDwlPw5F6uNYUlhWp3OX3CHpycH-MLOOvbhE_g&e=>. Can you and @mhlswang<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_mhlswang&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=ILbuN39K86TRVV4M0fXV7A&m=WTxe_jLFiZc19hQ0WBjz01ksdtPAwYEvCZEUs92JSJ0&s=YNnWk4oQTrBsyKUxyAqSaEdIJu-3pb94DVGauBVC5aI&e=> take a look to determine if that is an adequate solution?

Thank you @knoepfel<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_knoepfel&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=ILbuN39K86TRVV4M0fXV7A&m=WTxe_jLFiZc19hQ0WBjz01ksdtPAwYEvCZEUs92JSJ0&s=mIwOS1sM9GxtszUj9zSmTGErfrTEAWE7WgFwOm8iiyQ&e=> for your comments. I looked into NumCpp and it does not seem trivial to convert the code to use NumCpp. Mike and I had a little discussion on this. Even though the code is potentially useful to multiple experiments, we understand your concern with maintaining
c2numpy. Therefore, if you agree, we would like to delete this pull request. Experiments that wish to use this code can implement it in their own repository. Thanks.

**Comment by knoepfel** [comment]
@aihimmel and @yangtj207, thank you for your responses.  Tingjun, I don't think the SciSoft team is ready to abandon this PR yet, but I appreciate your willingness to heed our concerns.  See below.

Alex, much of the code in LArSoft exists because it is "potentially useful," and not because it actually is used.  The bar for inclusion now is higher than that, especially whenever you take into account the level of support available for the project.  Because `c2numpy` is not being maintained by an external party, we know *from the outset* that we are incurring a maintenance liability exceeding that of an externally maintained, third-party product.  Consequently, we should not rush it into the distribution.  That said, you are correct that we don't want multiple experiments coming up with their own solutions.

As the module under consideration would be used by perhaps more than one experiment, @yangtj207 and @mhlswang, I would be happy to meet with you to figure out a way to move this PR forward in a more maintainable way.

**Comment by yangtj207** [comment]
Hi @knoepfel what is your thought on this PR? We are still using this module a lot and it would be more convenient to have it in a release. Thanks.

**Comment by mhlswang** [comment]
@yangtj207, I had volunteered to find a suitable replacement for c2numpy which I had initially implemented in this code.  I hope to look into this soon.

**Comment by mhlswang** [comment]
I looked at the suggested alternative to c2numpy.h -- NpyCpp:

https://pmontalb.github.io/NpyCpp/ 

which seems to have been forked from cnpy:

https://github.com/rogersce/cnpy/

What we need to be able to do in our module is dump out the adc values of the raw waveforms, together with other relevant information, like pdgid, channel number, number of electrons deposited, etc. into a .npy file in the form of a structured numpy array where the dtypes of the columns are not of a uniform type.  Unfortunately NpyCpp appears to lack the flexibility to do this.  It looks like it can only dump out an array where all elements are of a uniform type.

On the other hand, c2numpy does exactly what we need by allowing one to define a structured array as you would in numpy with different dtypes for each column.  Could you remind me what the argument is against using c2numpy.  I looked at the code in some detail and it seems very straightforward with everything contained in a single header file.  I think it would be a lot simpler to use c2numpy.h or our own fork of it and maintain it ourselves -- which does not seem difficult to do.

**Comment by knoepfel** [comment]
Thank you for investigating, Mike.  We will include the `c2numpy.h` header file as part of the `larrecodnn` package.

**Comment by lgarren** [comment]
trigger build

**Comment by lgarren** [comment]
CI tests are normal

**Comment by lgarren** [comment]
approve

**Pull request merged**